### PR TITLE
Implement Groupby

### LINKED ISSUE #5: Implement Groupby or Aggregate Functions
For Group by, for each column A in relation R that appears in the SELECT clause, one of the following conditions must hold:
- A appears in the GROUP BY clause
- The primary key or R appears in the GROUP BY clause

List of Aggregate Functions:
- Min
- Max
- Count
- Avg

### PR DESCRIPTION
**Pull request by andlimey**

### CODE REVIEW

File: src/qp/optimizer/RandomInitialPlan.java [hunk 1]
```diff
@@ -195,10 +196,39 @@ public void createProjectOp() {
...
+        }
+    }
+
     public void createOrderbyOp() {
         Operator base = root;
         if (!orderbylist.isEmpty()) {
-            root = new Orderby(base, orderbylist, OpType.ORDERBY, isDesc);
+            if (!groupbylist.isEmpty()) {
+                // Combines groupby list and orderby list to maintain the groups when sorting.
+                ArrayList<Attribute> combinedList = new ArrayList<>(groupbylist);
```
Comment (lingzhiyu) [inline src/qp/optimizer/RandomInitialPlan.java:219]: does this combination lead to the expected results? For a query that has both groupby and orderby, is this implied that these operations can be composed?

File: src/qp/optimizer/RandomInitialPlan.java [hunk 2]
```diff
@@ -195,10 +196,39 @@ public void createProjectOp() {
...
+
+                // Doesn't include duplicated attributes in the combined list.
+                for (Attribute attr: orderbylist) {
+                    if (!combinedList.contains(attr)) {
+                        combinedList.add(attr);
+                    }
```
Comment (lingzhiyu) [inline src/qp/optimizer/RandomInitialPlan.java:225]: I'm assuming that the order of these attributes affects the expected result?

**Review by lingzhiyu — APPROVED** [review]
LGTM. Didn't look through much of the code that's for sorting.

File: src/qp/optimizer/RandomInitialPlan.java (hunk 2 above)
Comment (andlimey) [reply src/qp/optimizer/RandomInitialPlan.java:225]: Yes

**Pull request merged**